evm/utils/test_builder.py

```diff
@@ -0,0 +1,309 @@
+
+
+DEFAULT_ENV = OrderedDict([
+    ("currentCoinbase", decode_hex("0x2adc25665018aa1fe0e6bc666dac8fc2697ff9ba")),
+    ("currentDifficulty", 131072),
+    ("currentGasLimit", 1000000),
+    ("currentNumber", 1),
+    ("currentTimestamp", 1000),
+    ("previousHash", decode_hex(
+        "0x5e20a0453cecd065ea59c37ac63e079ee08998b6045136a8ce6635c7912ec0b6"
+    )),
+])
+
+ALL_FORKS = [
+    "Frontier",
+    "Homestead",
+    "EIP150",
+    "EIP158",
+    "Byzantium"
+]
+
+FILLED_WITH_TEMPLATE = "py-evm-{version}"
+
+
+def normalize_state(state):
+    if isinstance(state, Iterable) and not isinstance(state, Mapping):
+        state = state_list_to_dict(state)
+    for address, account_state_dict in state.items():
+        account_state_dict.setdefault("balance", 0)
+        account_state_dict.setdefault("nonce", 0)
+        account_state_dict.setdefault("code", b"")
+        account_state_dict.setdefault("storage", OrderedDict())
+    return state
+
+
+def state_list_to_dict(state_list):
+    d = {}
+    for state_item in state_list:
+        d = assoc_in(d, state_item[:-1], state_item[-1])
+    return d
+
+
+def merge_states(state1, state2):
+    new_state = copy.deepcopy(state2)
+    for key, value in state1.items():
+        if key not in new_state:
+            new_state[key] = copy.deepcopy(value)
+        elif isinstance(value, Mapping) and isinstance(new_state[key], Mapping):
+            new_state[key] = merge_states(new_state[key], value)
+    return new_state
+
+
+def without_duplicates(l):
+    return list(OrderedDict.fromkeys(l))
+
+
+def flatten(l):
+    return list(itertools.chain.from_iterable(l))
+
+
+def int_to_hex(i):
+    return encode_hex(int_to_big_endian(i))
+
+
+def get_version_from_git():
+    version = subprocess.check_output(["git", "describe"]).strip()
+    return force_text(version)
+
+
+env_formatter = curried.apply_formatters_to_dict({
+    "currentCoinbase": to_checksum_address,
+    "currentDifficulty": int_to_hex,
+    "currentGasLimit": int_to_hex,
+    "currentNumber": int_to_hex,
+    "currentTimestamp": int_to_hex,
+    "previousHash": encode_hex,
+})
+
+storage_item_formatter = combine_argument_formatters(int_to_hex, int_to_hex)
+
+
+def storage_formatter(storage_dict):
+    items = list(storage_dict.items())
+    formatted_items = apply_formatter_to_array(storage_item_formatter, items)
+    return dict(formatted_items)
+
+
+account_state_formatter = curried.apply_formatters_to_dict({
+    "balance": int_to_hex,
+    "nonce": int_to_hex,
+    "code": encode_hex,
+    "storage": storage_formatter,
+})
+
+
+state_item_formatter = combine_argument_formatters(to_checksum_address, account_state_formatter)
+
+
+def state_formatter(state_dict):
+    items = list(state_dict.items())
+    formatted_items = apply_formatter_to_array(state_item_formatter, items)
+    return dict(formatted_items)
+
+
+class StateTest:
+
+    def __init__(
+        self,
+        name,
+        *,
```

Comment:
I'm not familiar with this pattern.  What's it do?

Reply:
It requires the following arguments to be given by keyword (https://www.python.org/dev/peps/pep-3102/). This has the side effect that you can add parameters with no default after other keyword arguments (`transaction` and `secret_key`).

Reply:
I love this so much.  I'd like to backfill this pattern onto a bunch of places in py-evm and also the p2p code (cc @gsalgado).